src/platform/ModulesManager.c

```diff
@@ -0,0 +1,742 @@
+                        }
+                        else if (NULL == (reported[i].object = strdup(reported[i].object)))
+                        {
+                            OsConfigLogError(GetPlatformLog(), "Failed to allocate memory for object name");
+                        }
+                        else
+                        {
+                            if (IsFullLoggingEnabled())
+                            {
+                                OsConfigLogInfo(GetPlatformLog(), "Found reported property (%s.%s)", reported[i].component, reported[i].object);
+                            }
+
+                            reportedTotal++;
+                        }
+                    }
+
+                    OsConfigLogInfo(GetPlatformLog(), "Found %d reported objects in '%s'", reportedTotal, configJson);
+
+                    g_reported = reported;
+                    g_reportedTotal = reportedTotal;
+                }
+                else
+                {
+                    OsConfigLogError(GetPlatformLog(), "Failed to allocate memory for reported objects");
+                }
+            }
+        }
+    }
+
+    if (config)
+    {
+        json_value_free(config);
+    }
+}
+
+void AreModulesLoadedAndLoadIfNot(const char* directory, const char* configJson)
+{
+    if (NULL == g_modules)
+    {
+        LoadModules(directory, configJson);
+    }
+}
+
+static void FreeModules(MODULE* modules)
+{
+    MODULE* curr = modules;
+    MODULE* next = NULL;
+
+    while (curr)
+    {
+        next = curr->next;
+        UnloadModule(curr);
+        curr = next;
+    }
+
+    modules = NULL;
+}
+
+static void FreeSessions(SESSION* sessions)
+{
+    SESSION* curr = sessions;
+    SESSION* next = NULL;
+    MODULE_SESSION* moduleSession = NULL;
+
+    while (curr)
+    {
+        next = curr->next;
+
+        while (curr->modules)
+        {
+            moduleSession = curr->modules;
+            curr->modules = moduleSession->next;
+            FREE_MEMORY(moduleSession);
+        }
+
+        FREE_MEMORY(curr->uuid);
+        FREE_MEMORY(curr->client);
+        FREE_MEMORY(curr);
+
+        curr = next;
+    }
+
+    sessions = NULL;
+}
+
+static void FreeReportedObjects(REPORTED_OBJECT* reportedObjects, int numReportedObjects)
+{
+    int i = 0;
+
+    for (; i < numReportedObjects; i++)
+    {
+        FREE_MEMORY(reportedObjects[i].component);
+        FREE_MEMORY(reportedObjects[i].object);
+    }
+
+    FREE_MEMORY(reportedObjects);
+}
+
+void UnloadModules(void)
+{
+    FreeModules(g_modules);
+    FreeSessions(g_sessions);
+    FreeReportedObjects(g_reported, g_reportedTotal);
+
+    g_reportedTotal = 0;
+}
+
+static char* GenerateUuid(void)
+{
+    char* uuid = NULL;
+    static const char uuidTemplate[] = "xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx";
+    const char* hex = "0123456789ABCDEF-";
+    int random = 0;
+    char c = ' ';
+    int i = 0;
+
+    if (NULL == (uuid = (char*)malloc(UUID_LENGTH + 1)))
```

Comment:
>if (NULL == (uuid = (char*)malloc(UUID_LENGTH + 1))) [](http://example.com/codeflow?start=4&length=52)

memset all to 0 to avoid need to init each to 0 later